### PR TITLE
Specification constants support

### PR DESCRIPTION
**Pull request by pixelflinger**

### CODE REVIEW

File: filament/backend/include/backend/Program.h [hunk 1]
```diff
@@ -84,6 +84,15 @@ class Program {
     Program& setSamplerGroup(size_t bindingPoint, ShaderStageFlags stageFlags,
             Sampler const* samplers, size_t count) noexcept;
 
+    struct SpecializationConstant {
+        uint32_t id;                                // id set in glsl, needed for vulkan
```
Comment (bejado) [inline filament/backend/include/backend/Program.h:88]: both Vulkan and Metal yes?

File: filament/backend/src/metal/MetalHandles.mm [hunk 2]
```diff
@@ -372,7 +372,22 @@ BufferDescriptor b(nullptr, 0u, [](void* buffer, size_t size, void* user) {
...
+            const std::array<MTLDataType, 3> types{
+                MTLDataTypeInt, MTLDataTypeFloat, MTLDataTypeBool };
+            std::visit([&sc, constants, type = types[sc.value.index()]](auto&& arg) {
+                [constants setConstantValue:&arg
+                                       type:type
+                                    atIndex:sc.id];
+            }, sc.value);
+        }
+
+        *shaderFunctions[i] = [library newFunctionWithName:@"main0"
```
Comment (bejado) [inline filament/backend/src/metal/MetalHandles.mm:387]: nit: align the `:` on each line

File: filament/backend/include/backend/Program.h (hunk 1 above)
Comment (pixelflinger) [reply filament/backend/include/backend/Program.h]: yes thanks

**Review by prideout — APPROVED** [review]
Looks great.  Our only usage of std::variant, I think.

**Pull request merged**